fix for LocationParseError in netgear platform

etsinko opened the pull request:
## Description:

This is a proposed fix for 'LocationParseError' that happens whenever a Netgear router is auto-discovered. This error was caused by the fact that auto discovery platform does not pass the required configuration parameters (host, username, password and port) to netgear's `get_scanner` and therefore they all have `None` as their default values.

This PR fixes the issue by removing `netgear` platform from auto-discovered platforms. My argument for this is that auto-discovery routine does not provide host, username, password or port number. Even if some reasonable default values were used for cases like this the likelihood that users use the same default values for their routers is low.

**Related issue (if applicable):** fixes #4318 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable): N/A

## Checklist:

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] ~~Tests have been added to verify that the new code works.~~

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

etsinko committed:
fix for LocationParseError in netgear platform

etsinko committed:
added unit tests for get_scanner()

etsinko committed:
fixed houndci-bot warnings

etsinko committed:
fixed lint warnings

etsinko committed:
fixed lint warnings

etsinko committed:
fixed broken test

balloob commented on homeassistant/components/device_tracker/netgear.py:
```diff
@@ -37,6 +37,9 @@ def get_scanner(hass, config):
     password = info.get(CONF_PASSWORD)
     port = info.get(CONF_PORT)
+    # Check that all parameters are provided
+    if not all([host, username, password, port]):
```
Don't do this. The platform schema already validates it.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -89,9 +88,7 @@
-DISCOVERY_PLATFORMS = {
-    SERVICE_NETGEAR: 'netgear',
-}
+DISCOVERY_PLATFORMS = {}
```
Please remove all discovery related code from `device_tracker`.

etsinko committed:
removed guard clause from netgear.py
removed all discovery related code from device_tracker
removed unnecessary unit test

etsinko committed:
removed discovery related tests

etsinko committed:
removed unused import

etsinko committed:
removed unused import

etsinko commented:
@balloob it is done. All discovery related code has been removed from `device_tracker`.

MartinHjelmare commented:
@balloob Removing all discovery related code has broken the device tracker for some platforms, at least mysensors and looks like Volvo too.

https://github.com/home-assistant/home-assistant/issues/10124

MartinHjelmare commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -180,22 +175,6 @@ def async_setup_platform(p_type, p_config, disc_info=None):
-        hass, DISCOVERY_PLATFORMS.keys(), async_device_tracker_discovered)
-
-    @asyncio.coroutine
-    def async_platform_discovered(platform, info):
```
This is still needed.

etsinko replied:
@MartinHjelmare I think ```async_device_tracker_discovered()``` is not needed, because it was only used by `netgear` platform. What should be returned is `async_platform_discovered()`

MartinHjelmare replied:
Yes, only add the code that starts on the lines where my comment starts, eg here that's 93. I haven't marked the import that's needed btw.

MartinHjelmare commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -180,22 +175,6 @@ def async_setup_platform(p_type, p_config, disc_info=None):
-        """Load a platform."""
-        yield from async_setup_platform(platform, {}, disc_info=info)
-
-    discovery.async_listen_platform(hass, DOMAIN, async_platform_discovered)
```
This is still needed for general platform discovery to work.

MartinHjelmare commented on tests/components/device_tracker/test_init.py:
```diff
@@ -311,36 +310,6 @@ def test_mac_vendor_lookup_on_see(self):
-    @patch(
-        'homeassistant.components.device_tracker.demo.setup_scanner',
-        autospec=True)
-    def test_discover_platform(self, mock_demo_setup_scanner, mock_see):
```
This test should be kept.

etsinko commented:
@MartinHjelmare Is there a way to check that returning discovery related code in fact fixes the reported problems?

MartinHjelmare reviewed (commented):
The code below should be put back.

MartinHjelmare commented:
@etsinko platforms that depend on discovery info when setting up the device tracker platform will not work without the marked code above.

etsinko commented:
I wonder if I revert these faulty commits in the same branch will it show up in this PR?

MartinHjelmare commented:
Not as long as this PR is closed. When you open the new PR the branch should be tracked there.

etsinko commented:
@MartinHjelmare please see PR #10169
